frame/sudo/src/lib.rs

```diff
@@ -214,3 +216,47 @@ decl_error! {
 		RequireSudo,
 	}
 }
+
+/// Simple pass through for the weight function, but fee isn't paid because it is Sudo.
+struct SudoPassthrough<Call>(sp_std::marker::PhantomData<Call>);
+
+impl<Call> SudoPassthrough<Call> {
+	fn new() -> Self { Self(Default::default()) }
+}
+impl<Call: GetDispatchInfo> WeighData<(&Box<Call>,)> for SudoPassthrough<Call> {
+	fn weigh_data(&self, (call,): (&Box<Call>,)) -> Weight {
+		call.get_dispatch_info().weight + 10_000
+	}
+}
+impl<Call: GetDispatchInfo> ClassifyDispatch<(&Box<Call>,)> for SudoPassthrough<Call> {
+	fn classify_dispatch(&self, (call,): (&Box<Call>,)) -> DispatchClass {
+		call.get_dispatch_info().class
+	}
+}
+impl<Call: GetDispatchInfo> PaysFee<(&Box<Call>,)> for SudoPassthrough<Call> {
+	fn pays_fee(&self, (call,): (&Box<Call>,)) -> bool {
+		call.get_dispatch_info().pays_fee
```

Comment:
Shouldn't this be true?

Reply:
The way it is written, user only pays a fee if the underlying call would pay a fee. If someone called a "free" transaction, it would also be free.

Sounds like the right behavior to me?

Reply:
We just had the "problem" that we can not have calls that don't need to pay any fees. If we have such calls, it opens a ddos door. You could spam the system from accounts without any funds. The calls would be valid, but would fail at applying because the users aren't root. As @xlc said, this is a more complex topic and would require a `SignedExtension` to pay the fee back, if the origin is the sudo account.

Reply:
@bkchr see the implementation in Util https://github.com/paritytech/substrate/blob/master/frame/utility/src/lib.rs#L192

Should that also then be updated? It seems the issue here is not with this part of the logic, but including extrinsics which would pay a fee.

Reply:
So, the system is currently not perfect. However, you could have the following:

- You have transaction `A` that doesn't pay any fees, but you protect this with a custom `SignedExtension`. This `SignedExtension` makes sure that `validate_transaction` fails when the origin for example is not correct.
- With the current propagation of `pays_fee`, a malicious user could create a `Sudo` transaction that wraps the `A` transaction and it would be free. The malicious user could start to spam the network with this `Sudo(A)` transaction.
- `SignedExtension` currently don't propagate the checks if your call wraps another call, which could you bring into this situation. So, the best for the moment is just to always pay fees.